src/coreclr/vm/interoplibinterface_shared.cpp

```diff
@@ -154,3 +154,138 @@ void Interop::OnAfterGCScanRoots(_In_ bool isConcurrent)
         ObjCMarshalNative::AfterRefCountedHandleCallbacks();
 #endif // FEATURE_OBJCMARSHAL
 }
+
+#ifdef FEATURE_GCBRIDGE
+
+namespace
+{
+    Volatile<BOOL> g_GCBridgeActive = FALSE;
+    CLREvent* g_bridgeFinished = nullptr;
+
+    void ReleaseGCBridgeArgumentsWorker(
+        _In_ size_t sccsLen,
+        _In_ StronglyConnectedComponent* sccs,
+        _In_ ComponentCrossReference* ccrs)
+    {
+        CONTRACTL
+        {
+            NOTHROW;
+            GC_NOTRIGGER;
+        }
+        CONTRACTL_END;
+
+        // Memory was allocated for the collections by the GC.
+        // See callers of GCToEEInterface::TriggerGCBridge().
+
+        // Free memory in each of the SCCs
+        for (size_t i = 0; i < sccsLen; i++)
+        {
+            free(sccs[i].Context);
+        }
+        free(sccs);
+        free(ccrs);
+    }
+}
+
+bool Interop::IsGCBridgeActive()
+{
+    LIMITED_METHOD_CONTRACT;
+
+    return g_GCBridgeActive;
+}
+
+void Interop::WaitForGCBridgeFinish()
+{
+    CONTRACTL
+    {
+        MODE_COOPERATIVE;
+    }
+    CONTRACTL_END;
+
+    while (g_GCBridgeActive)
+    {
+        GCX_PREEMP();
+        g_bridgeFinished->Wait(INFINITE, false);
+        // In theory, even though we waited for bridge to finish, because we are in preemptive mode
+        // the thread could have been suspended and another GC could have happened, triggering bridge
+        // processing again. In this case we would wait again for bridge processing.
+    }
+}
+
+void Interop::TriggerClientBridgeProcessing(
+    _In_ size_t sccsLen,
+    _In_ StronglyConnectedComponent* sccs,
+    _In_ size_t ccrsLen,
+    _In_ ComponentCrossReference* ccrs)
+{
+    CONTRACTL
+    {
+        NOTHROW;
+        GC_NOTRIGGER;
+    }
+    CONTRACTL_END;
```

Comment:
Can we add an a `PRECONDITION` for this being called during GC suspension? Makes it easier to understand the flow without tracing all of the callers (same for JavaNative::TriggerClientBridgeProcessing) especially since the contract doesn't have a GC mode requirement.